### PR TITLE
Avoid ClassCastException when getting values from `QueryContext`

### PR DESCRIPTION
**Pull request by FrankChen021**
The problem was first reported by #12760 , and #12833 solves that specific issue.
This PR solves all potential problems by providing a group of type safe methods.

### Description

`QueryContext` holds values passed from user side in JSON format. 
And currently Druid requires values in JSON must be strictly type matched. 
For example, a property `maxOnDiskStorage` in Druid is defined as integer, it must be in the format as 

```json
"context": {"maxOnDiskStorage": 100} 
```

If the number 100 is serialized as string as 

```json
"context": {"maxOnDiskStorage": "100"} 
```

Druid throws a `ClassCastException` to reject the input query.

Actually, string-formatted number "100" can be parsed as number at Druid server side. 

### Key changes

#12833 adds a method `getContextHumanReadableBytes` to get context value from any acceptable format safely. And this PR adds a group of methods to do so and changes all existing code reference to the new APIs.

Newly added methods are:

```java
getContextAsInt
getContextAsLong
getContextAsFloat
getContextAsString
getContextAsEnum
getContextAsBoolean
```

Original unsafe method `getContextValue` is still kept to allow us get objects which are stored by Druid server side temporarily.
Some javadoc is added to tell callers know which methods should be used instead.

This PR has:
- [X] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [X] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [X] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

**Comment by FrankChen021** [comment]
@clintropolis Could you review this?

**Review by clintropolis — APPROVED** [review]
seems reasonable to me :+1:

File: processing/src/main/java/org/apache/druid/query/Query.java [hunk 1]
```diff
@@ -124,22 +124,130 @@ default QueryContext getQueryContext()
     return null;
   }
 
-  <ContextType> ContextType getContextValue(String key);
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/Query.java:127]: hmm, actually I guess this is an `@ExtensionPoint`, we should probably leave the old methods here to not be backwards incompatible

File: processing/src/main/java/org/apache/druid/query/QueryContexts.java [hunk 2]
```diff
@@ -604,18 +644,31 @@ private QueryContexts()
   {
   }
 
-  static <T, E extends Enum<E>> E parseEnum(Query<T> query, String key, Class<E> clazz, E defaultValue)
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/QueryContexts.java:607]: this class is also marked `@PublicApi` and so we should ideally leave the signatures in place

File: processing/src/main/java/org/apache/druid/query/Query.java [hunk 3]
```diff
@@ -124,22 +124,130 @@ default QueryContext getQueryContext()
...
+    if (getQueryContext() != null) {
+      return getQueryContext().getAsInt(key);
+    } else {
+      return null;
+    }
+  }
+
+  default int getContextAsInt(String key, int defaultValue)
+  {
+    if (getQueryContext() != null) {
```
Comment (paul-rogers) [inline processing/src/main/java/org/apache/druid/query/Query.java:178]: Nit: use `==` and revers the then and `else` clauses. Makes the code just a bit easier to read.

File: processing/src/main/java/org/apache/druid/query/Query.java [hunk 4]
```diff
@@ -124,22 +124,130 @@ default QueryContext getQueryContext()
...
+      return getQueryContext().getAsEnum(key, clazz, defaultValue);
+    } else {
+      return defaultValue;
+    }
+  }
+
+  @Nullable
+  default Boolean getContextAsBoolean(String key)
```
Comment (paul-rogers) [inline processing/src/main/java/org/apache/druid/query/Query.java:223]: Nit: use the same name for both "flavors" of methods: easier to remember.

File: processing/src/main/java/org/apache/druid/query/QueryContext.java [hunk 5]
```diff
@@ -168,7 +168,20 @@ public Object getOrDefault(String key, Object defaultValue)
...
+  }
+
+  public String getAsString(String key, String defaultValue)
+  {
+    Object val = get(key);
+    return val == null ? defaultValue : val.toString();
+  }
+
+  @Nullable
+  public Boolean getAsBoolean(String parameter)
```
Comment (paul-rogers) [inline processing/src/main/java/org/apache/druid/query/QueryContext.java:182]: Nit: best to not call query context values "parameters" since a query can have "parameters" which are not in the context. Suggestion: "key".

File: processing/src/main/java/org/apache/druid/query/QueryContext.java [hunk 6]
```diff
@@ -179,6 +192,11 @@ public boolean getAsBoolean(
     return QueryContexts.getAsBoolean(parameter, get(parameter), defaultValue);
   }
 
+  public Integer getAsInt(final String parameter)
```
Comment (paul-rogers) [inline processing/src/main/java/org/apache/druid/query/QueryContext.java:195]: I wonder, since we have methods for all of these on `QueryContexts`, and we use `QueryContexts` to get each specific value type, should we just omit these methods and standard on always using `QueryContexts` to get context values? Thoughts?

File: processing/src/main/java/org/apache/druid/query/QueryContexts.java [hunk 7]
```diff
@@ -399,12 +398,12 @@ public static <T> Query<T> verifyMaxQueryTimeout(Query<T> query, long maxQueryTi
...
   public static <T> long getMaxQueuedBytes(Query<T> query, long defaultValue)
   {
-    return parseLong(query, MAX_QUEUED_BYTES_KEY, defaultValue);
+    return query.getContextAsLong(MAX_QUEUED_BYTES_KEY, defaultValue);
   }
 
   public static <T> long getMaxScatterGatherBytes(Query<T> query)
   {
-    return parseLong(query, MAX_SCATTER_GATHER_BYTES_KEY, Long.MAX_VALUE);
+    return query.getContextAsLong(MAX_SCATTER_GATHER_BYTES_KEY, Long.MAX_VALUE);
```
Comment (paul-rogers) [inline processing/src/main/java/org/apache/druid/query/QueryContexts.java:406]: Thanks for cleaning this up! I'd wondered why we have two different ways to get values.

File: processing/src/main/java/org/apache/druid/query/QueryContexts.java [hunk 8]
```diff
@@ -503,7 +487,24 @@ public static String getAsString(
...
     } else {
-      throw new IAE("Expected parameter [%s] to be String", parameter);
+      throw new IAE("Expected parameter [%s] to be a String, but got [%s]", parameter, value.getClass().getName());
+    }
+  }
+
+  @Nullable
+  public static Boolean getAsBoolean(
+      final String parameter,
+      final Object value
```
Comment (paul-rogers) [inline processing/src/main/java/org/apache/druid/query/QueryContexts.java:497]: This set of methods has gotten a bit muddled.

* This version takes the name of a context key (not "parameter") just so it can throw an error. But, there are places where we (or at least I) want to use this method where the map is not a context.
* There are versions that take a map and a default, and versions, like this, that take a value (not parameter) and convert it.

Maybe have the following set:

```java
// Check if a key exists for code that wants to do something special for this case
public boolean hasValue(Map<..> context, String key) ...

// Generic method to convert objects to boolean
public boolean parseBoolean(Object value)
{
    // if can't parse, throw an exception
}

// Specific method for boolean context values
public boolean getBoolean(Map<...> context, String key, boolean defaultValue)
{
  value = context.get(key);
  if (value == null) {
    return defaultValue;
  }
  try {
      return parseBoolean(value);
   } catch (Exception e) {
     // Throw the "Expected context value [%s] to be a ... exception
  }
}
```

```

File: processing/src/main/java/org/apache/druid/query/QueryContexts.java [hunk 9]
```diff
@@ -604,18 +644,31 @@ private QueryContexts()
...
+    }
+    catch (IllegalArgumentException e) {
+      throw new IAE("Expected parameter [%s] must be value of enum [%s], but got [%s].",
+                    parameter,
+                    clazz.getName(),
+                    val.toString());
     }
+
+    throw new ISE(
+        "Expected parameter [%s] must be type of [%s], actual type is [%s].",
```
Comment (paul-rogers) [inline processing/src/main/java/org/apache/druid/query/QueryContexts.java:668]: Again, context values are not "parameters".

**Review by paul-rogers — COMMENTED** [review]
@FrankChen021, thanks for helping clean up this area! It is in much need of tidying up. A few comments for your consideration.

File: processing/src/main/java/org/apache/druid/query/QueryContexts.java (hunk 2 above)
Comment (FrankChen021) [reply processing/src/main/java/org/apache/druid/query/QueryContexts.java:607]: But this method is not declared as public. Do we need to keep it?

File: processing/src/main/java/org/apache/druid/query/Query.java (hunk 1 above)
Comment (FrankChen021) [reply processing/src/main/java/org/apache/druid/query/Query.java:127]: Let me revert this.

File: processing/src/main/java/org/apache/druid/query/Query.java (hunk 4 above)
Comment (FrankChen021) [reply processing/src/main/java/org/apache/druid/query/Query.java:223]: I think you mean this method `getContextBoolean`? 
I didn't change this method to use the same name pattern as 'getContextAs...' because there're lots reference to this method which might cause this PR so large. But indeed, it should be the same name pattern.

File: processing/src/main/java/org/apache/druid/query/QueryContexts.java (hunk 8 above)
Comment (FrankChen021) [reply processing/src/main/java/org/apache/druid/query/QueryContexts.java:497]: A `parseBoolean` is still kept at line 474 in this class to allow to parse a value from a map object.

**Comment by paul-rogers** [comment]
@FrankChen021, as it turns out, the `QueryContext` class introduced several problems. [PR 13049](https://github.com/apache/druid/pull/13049) proposes to rip out `QueryContext` and go back to using a map (along with a few adjustments to allow authorizing the context.) That PR will, unfortunately, conflict with this one.

While using maps solves some problems, it requires awkward code such as:

```java
x = QueryContexts.getSomething(context, key, default)
```

One advantage of the old `QueryContext` was that we could do:

```java
x = queryContext.getSomething(key, default);
```

Here, you propose a solution: add those values to the query class. That's OK, but it makes the query class pretty complex: we have methods for many tasks, and now we add multiple methods for context values. Plus, there are places (such as the SQL layer) where we work with the context without also having a native query. So, we'd have to duplicate these methods in several places. This suggests that methods to work with the context belong on the context itself, not on the users of the context (such as `Query`).

One solution, which I didn't include in PR 13049, is to define a new class: a typed wrapper on top of the map. Something like:

```java
class QueryContext {
  Object get(String key);
  String getString(String key);
  String getString(String key, String defaultValue);
  int getInt(String key);
  int getInt(String key, int defaultValue);
  ...
}
```

For each of our supported types: String, Int, Long, etc. We'd then store an instance of `QueryContext` where we currently store the map.

`QueryContexts` has a zillion `getSomething()` methods for the various parameters. The logical step would be to move those into the above `QueryContext` class, else we'll find ourselves either duplicating code, or still using `QueryContexts` methods as shown earlier.

The problem that PR 13049 tries to solve is that there are times when the context is immutable, other times when it is mutable. And, those times trade off: the user's request is immutable, a mutable copy is made while planning the query, then the context must become immutable again once we start executing in multiple threads. So, we could have two forms: the above immutable form with only "getters", and a mutable form:

```java
class MutableQueryContext {
  Object put(String key, Object value);
  Object remove(String key);
  ...
}
```

To do this, we'd actually want `QueryContext` to be an interface, with two concrete implementations: `ImmutableQueryContext` and `MutableQueryContext`. One reason for the two classes is that the classes will clearly label the intent of when when the context is mutable or not.

This may still lead to a muddle, however. A `Query` will have a mutable context during the "planning" (`QueryRunner`-based) part of the native query execution, but immutable once we move to the execution (`Sequence`-based) part of the process.

There is also the issue of compatibility with custom `Query` classes where folks have implemented the existing interfaces using a map. It is not entirely clear how often that is done and what we can change in this area. Any ideas?

**Comment by paul-rogers** [comment]
Here's a follow-on idea. Modifying a context map can perhaps be left to working directly with the map, since it is mostly just calling `put()` along with a few other items. So, perhaps we can just focus on reading the context. Maybe we define:

```java
interface Query<...> {
   default QueryContext getQueryContext() { return new QueryContext(getContext()); }
```

Then, `QueryContext` has all the many accessor methods mentioned above. That is, `QueryContext` is a transient facade on top of the map.

With this approach, all the myriad `getFoo` and `parseFoo` method in `QueryContexts` can move to `QueryContext`.

What do you think? If this will work, I can perhaps modify the PR I mentioned to include this approach.

**Comment by paul-rogers** [comment]
Went ahead and added the `QueryContext` "facade" in PR #13049 as it is a cleaner solution to the race condition issue which that PR seeks to solve.

**Comment by FrankChen021** [comment]
> Here's a follow-on idea. Modifying a context map can perhaps be left to working directly with the map, since it is mostly just calling `put()` along with a few other items. So, perhaps we can just focus on reading the context. Maybe we define:
> 
> ```java
> interface Query<...> {
>    default QueryContext getQueryContext() { return new QueryContext(getContext()); }
> ```
> 
> Then, `QueryContext` has all the many accessor methods mentioned above. That is, `QueryContext` is a transient facade on top of the map.
> 
> With this approach, all the myriad `getFoo` and `parseFoo` methods in `QueryContexts` can move to `QueryContext`. In this approach, the goal of this PR (type safety) would be provided by the typed `get{Type}` methods on `QueryContext`.
> 
> What do you think? If this will work, I can perhaps modify the PR I mentioned to include this approach.

Hi @paul-rogers 

When I touch this part in this PR, I noticed there are two classes, `QueryContext` and `QueryContexts`, the latter one provides a group by getXXXX helpers that QueryContext can be used. And also, another group of `getContextXXXX` methods are defined in `Query` to get values from `QueryContext`. 
As you can seen in this PR, if a new `getXXX` method is added, we have to add it in 3 classes to follow current pattern. These are too complicated. I didn't make changes to this part because it involves a huge refactoring.

Basically, I agree with you that there should be a facade `QueryContext` there(but I think QueryContext should not expose `Map` object to callers directly), and a series of `getXXXX` methods are defined in `QueryContext` to get values. And at last those helper motheds in `Query` are removed or deprecated.

My focus in this PR is on the safe type conversion at caller side. And the work has been done. How about let's merge this first and then resolve the conflicts after you stablize the core changes on your branch?

**Comment by paul-rogers** [comment]
@FrankChen021, your goal of type-safe conversion is a good one. The question is how to achieve that goal so that you can get this PR done, while also allowing us to solve the other issues in the other PR.

Perhaps the most expedient solution is the one you suggest: go ahead and commit this PR. The other one will undo the new `Query` methods; but that should be OK because they won't have been around long enough for extensions to use them. Replacing calls to your new methods with the other approach should just be a simple text replace operation, which I can do in the other PR.

The question we're asking is: should we add more methods to the `Query` interface, or should we move those responsibilities to a new class? In this PR, we've added more `getContextX` methods to `Query`. In the other PR, we've actually deprecated the existing ones in favor of the new `QueryContext` facade.

If we want to consider alternatives, let's note that if we commit this one first, then the other PR (assuming it is approved), would immediately remove the new `getContextX` methods: they'd live only for the time between the two commits. So, perhaps an alternative solution is to use the existing `QueryContexts` methods in this PR, then switch to the facade in the next PR. For example, consider this line in this PR:

```java
final String timestampStringFromContext = query.getContextAsString(CTX_KEY_FUDGE_TIMESTAMP, "");
```

With the suggested "temporary" solution, perhaps write this as:

```java
final String timestampStringFromContext = QueryContexts.getAsString(query, CTX_KEY_FUDGE_TIMESTAMP, "");
```

In the later PR, this would become:

```java
final String timestampStringFromContext = query.queryContext().getString(CTX_KEY_FUDGE_TIMESTAMP, "");
```

The alternative approach avoids adding new methods to the `Query` interface.

Otherwise, the two approaches are basically the same. There is one other difference, however: all the `getFoo()` methods in `QueryContexts`. With the approach in this PR, we'd have two ways to get values:

```java
final String timestampStringFromContext = query.getContextAsString(CTX_KEY_FUDGE_TIMESTAMP, "");
final boolean flag = QueryContexts.getEnableJoinFilterRewrite(query);
```

In the other PR, we can move all the getters to a single class:

```java
final QueryContext queryContext = query.queryContext();
final String timestampStringFromContext = queryContext.getString(CTX_KEY_FUDGE_TIMESTAMP, "");
final boolean flag = queryContext.enableJoinFilterRewrite();
```

The result is simpler code, while providing type safety, concurrency safety, and the ability to authorize context values.

So, the only question is how we help you get this one done now, and how we move onto the broader fix a bit later.

Thoughts?

**Comment by FrankChen021** [comment]
@paul-rogers I think the new APIs you proposed is what I want -- that is there's only one place for these `getXXX` methods.

```java
final QueryContext queryContext = query.queryContext();
final String timestampStringFromContext = queryContext.getString(CTX_KEY_FUDGE_TIMESTAMP, "");
final boolean flag = queryContext.enableJoinFilterRewrite();
```

Let me remove all newly addde `getContextXXXX` methods in this PR from `Query` interface, and replace all reference to these methods to what you suggest as 

```java
final String timestampStringFromContext = QueryContexts.getAsString(query, CTX_KEY_FUDGE_TIMESTAMP, "");
```

I will do this ASAP so that it won't block other PRs for long time.

**Comment by FrankChen021** [comment]
@paul-rogers I've replace all reference from calling `getContextAsXXXX` to `getQueryContext().getAsXXXX`, which is more similar to the final APIs that you propose.

old one:
```
final String timestampStringFromContext = query.getContextAsString(CTX_KEY_FUDGE_TIMESTAMP, "");
```

new one:
```
final String timestampStringFromContext = query.getQueryContext().getAsString(CTX_KEY_FUDGE_TIMESTAMP, "");
```

**Comment by FrankChen021** [comment]
@paul-rogers Is there anything that I need to do in this PR? If it looks good to you, I think we can merge it to continue #13071

**Review by paul-rogers — COMMENTED** [review]
LGTM (non-binding)

**Comment by FrankChen021** [comment]
Thank you @paul-rogers @clintropolis 
I will merge this.

**Comment by clintropolis** [comment]
the last commit of this PR didn't actually run through travis and seems to be failing https://github.com/apache/druid/runs/8325253980, looking into the issue

**Comment by clintropolis** [comment]
opened #13083 to fix the test

**Comment by FrankChen021** [comment]
@clintropolis Thanks for the fix. I didn't notice that travis not started but just saw the CI status of the last commit is OK. So why did the travis not start?
